dimod/include/dimod/quadratic_model.h

```diff
@@ -873,6 +609,7 @@ class BinaryQuadraticModel : public QuadraticModelBase<Bias, Index> {
     /// Add quadratic bias for the given variables.
     void add_quadratic(index_type u, index_type v, bias_type bias) {
         if (u == v) {
+            assert(0 <= u && static_cast<size_t>(u) < this->num_variables());
```

Comment:
Shouldn't you want to keep these checks in release version as well (not just debug)?

Reply:
This is following C++'s style of allowing undefined behaviour in the interests of speed. So at the C++ level we allow segfaults. We do check this stuff at the Python level Python doesn't allow segfaults.

Reply:
If we'll stick to allowing segfaults in c++, once we stabilize/open c++ interface for devs, we'll need to heavily document that.

Reply:
Totally agree. I intend to take a pass through and mark the undefined behavior and/or exception safety appropriately. Just focusing on the functionality for the moment.